### PR TITLE
Stashcubelist

### PR DESCRIPTION
**Pull request by marqh**
usability tweak to `CubeList.__str__` to print a STASH code, if it exists and otherwise 
`unknown / (unknown)` 
would otherwise be returned

### CODE REVIEW

File: lib/iris/cube.py [hunk 1]
```diff
@@ -1835,6 +1835,10 @@ def summary(self, shorten=False, name_padding=35):
 
         nameunit = '{name} / ({units})'.format(name=self.name(),
                                                units=self.units)
+        # If all unknown and a STASH attribute exists, use it.
+        if (nameunit == 'unknown / (unknown)' and
+           'STASH' in self.attributes):
```
Comment (bjlittle) [inline lib/iris/cube.py]: This line can be joined to the above and it'll still only be 78 characters long.

File: lib/iris/cube.py [hunk 2]
```diff
@@ -1835,6 +1835,10 @@ def summary(self, shorten=False, name_padding=35):
...
+            nameunit = '{}'.format(self.attributes.get('STASH'))
```
Comment (bjlittle) [inline lib/iris/cube.py]: `self.attributes['STASH']` ... if you wanna, as you've already checked that the `STASH` key is available.

**Comment by marqh** [comment]
minor changes, as per review comments

File: lib/iris/cube.py [hunk 3]
```diff
@@ -1835,6 +1835,9 @@ def summary(self, shorten=False, name_padding=35):
 
         nameunit = '{name} / ({units})'.format(name=self.name(),
                                                units=self.units)
+        # If all unknown and a STASH attribute exists, use it.
+        if (nameunit == 'unknown / (unknown)' and 'STASH' in self.attributes):
```
Comment (ajdawson) [inline lib/iris/cube.py]: No need for parentheses when on one line.

**Comment by ajdawson** [comment]
Does this affect how `Cube`s are printed too? I'm wary of removing a consistency like this, users expect `name / (unit)`, and although this is not necessarily directly useful, it does flag that something is missing. This may not be the case if this is replaced by a stash code.

**Comment by marqh** [comment]
> Does this affect how Cubes are printed too? I'm wary of removing a consistency like this, users expect name / (unit), and although this is not necessarily directly useful, it does flag that something is missing. This may not be the case if this is replaced by a stash code.

this doesn't change cube print `cube.__str__`, but it does change `cube.__repr__`

I think this is a reasonable change, as most code will use print and this will still say `unknown / (unknown)` and then print the attributes, including STASH if available

the aim is to improve the printing of cube lists, where there is unknown name and unknown unit

**Comment by ajdawson** [comment]
OK great, thanks for clarifying @marqh.

**Comment by ajdawson** [comment]
One last thing then: did you consider `stash / (unknown)` and reject in favour of just the STASH? I can see both sides of the argument, the STASH is not the name, so it may not be appropriate to follow the `name / (unit)` convention, but on the other hand, we still can't give the units, so should we include that part?

**Comment by marqh** [comment]
> One last thing then: did you consider stash / (unknown) and reject in favour of just the STASH? I can see both sides of the argument, the STASH is not the name, so it may not be appropriate to follow the name / (unit) convention, but on the other hand, we still can't give the units, so should we include that part?

I did consider this; in principal we shouldn't have a name without a unit, so if we're falling back to STASH, we already don't have a unit

i do wonder whether `stashmsi / (unknown)` might be more specific, but it is longer, and i think it will always say unknown, as, if there is a unit, we'll print that rather than a STASH code

**Comment by ajdawson** [comment]
I guess the current strategy is fine, and really shines when someone knows what a STASH is and realises it probably tells them everything they need to know (or equips them with what they need to query with their data provider).

**Comment by bjlittle** [comment]
@marqh Did you consider something like `stash=m01s00i004` rather than just the bare stash code (i.e. `m01s00i004`) being used instead of `unknown / (unknown)`

**Comment by ajdawson** [comment]
It would certainly make it clear that the presented information was different somehow from the normal presentation of name and unit.

**Comment by marqh** [comment]
> Did you consider something like stash=m01s00i004 rather than just the bare stash code (i.e. m01s00i004) being used instead of unknown / (unknown)

I am content to do this, if you think it helps, though I don't like the `=` sign in the string

I'm keen to converge on one solution, so far we have

1. m01s00i004
1. m01s00i004 / (unknown)
1. stash: m01s00i004
1. stash=m01s00i004

please vote for your favorite or nominate another representation.  I'll code up the preferred solution

**Comment by marqh** [comment]
> please vote for your favorite or nominate another representation. I'll code up the preferred solution

@ajdawson @bjlittle ...

ty

**Comment by ajdawson** [comment]
I don't like 3, there are too many colons already in cubelist printouts. Number 2 is my next least favourite (I know, I brought this up...). I'd go with number 4, but if there are objections to this then my next choice is plain old number 1 (i.e. no change).

**Comment by bjlittle** [comment]
Apologies, I've been distracted ...

Okay, so I'm kinda in agreement with @ajdawson ... but I don't want this to be a thing, so old number 1 it is.

By golly I think we have a quorum!

**Pull request merged**